windowing/gbm: add EGL fencing for atomic drm

lrusak opened the pull request:
This is an alternative to #23793 and possibly #22012

This uses EGL fencing to allow async rendering via non-blocking atomic commits.

This is mostly taken from the kmscube implementation here -> https://gitlab.freedesktop.org/mesa/kmscube/-/commit/497ab787e5ab56c630014a20b43bfab16e3b43aa

This commit has been around since 2018 when I first implemented it. I never PR'd it because many platforms didn't use an up to date mesa or allowed for async drm.

This commit also adds a check that the DRM driver allows for async behaviour.

It seems to work ok on my AMD GPU. It would be nice to check on a lower power platform.

smp79 commented:
I did a quick test on Intel. This PR does offer some performance improvement but it's not even close to https://github.com/xbmc/xbmc/pull/23793 and  https://github.com/xbmc/xbmc/pull/22012

neo1973 reviewed (commented):
Looks correct as far as I can tell and performance is noticeably better with a CPU limited debug build :slightly_smiling_face: 

But what I'd like to understand is if the `WaitSyncGPU()` after the page flip is the most optimal solution. Doesn't this imply double buffering? With more then two buffers there would be no danger to render in the buffer that's currently scanned out, right?

smp79 commented:
Commenting out `WaitSyncGPU()` makes the GUI fps skyrocket.

lrusak commented:
> Looks correct as far as I can tell and performance is noticeably better with a CPU limited debug build 🙂
> 
> But what I'd like to understand is if the `WaitSyncGPU()` after the page flip is the most optimal solution. Doesn't this imply double buffering? With more then two buffers there would be no danger to render in the buffer that's currently scanned out, right?

Yea, I think this was originally written when we only had double buffering. Now we can have multiple surface buffers. I added a test commit to try and add an additional fence, I'm not sure if that helps or not.

lrusak commented on xbmc/windowing/gbm/drm/DRMUtils.cpp:
```diff
@@ -421,6 +421,18 @@ bool CDRMUtils::InitDrm()
 #endif
 
+  uint64_t supported{0};
+  ret = drmGetCap(m_fd, DRM_CAP_ASYNC_PAGE_FLIP, &supported);
```
hmmm, I'll have to look into this.

lrusak replied:
Seems it's irrelevant so I dropped it. On i.MX6 `DRM_CAP_ASYNC_PAGE_FLIP` isn't supported but using `DRM_MODE_ATOMIC_NONBLOCK` still works fine.

smp79 commented:
Build failure -

> xbmc/windowing/gbm/WinSystemGbmGLESContext.cpp: In member function 'virtual void KODI::WINDOWING::GBM::CWinSystemGbmGLESContext::PresentRender(bool, bool)':
> /home/a/LibreELEC.tv/build.LibreELEC-Generic.x86_64-12.0-devel/build/kodi-64d1844655563c6f29f52b4b7308bd3455960ef6/xbmc/windowing/gbm/WinSystemGbmGLESContext.cpp:140:19: error: base operand of '->' has non-pointer type 'std::array<std::unique_ptr<KODI::UTILS::EGL::CEGLFence>, 2>'
>   140 |         m_eglFence->CreateGPUFence();

neo1973 commented:
> Commenting out `m_eglFence->WaitSyncGPU()` makes the GUI fps skyrocket.

Skyrocket like "performance is greatly improved" or "something is broken"?

smp79 commented:
Greatly improved. Like, 45 fps vs 25 fps.

smp79 commented:
Without WaitSyncGPU()
![IMG_20231023_163338](https://github.com/xbmc/xbmc/assets/28512968/dbdf5c44-4c12-4351-b1c3-65af4fe9917f)

With WaitSyncGPU()
![IMG_20231023_163715](https://github.com/xbmc/xbmc/assets/28512968/86ca3c5b-ea8f-46d1-b7c3-e7676a054e14)

Tested with a 1080i50 stream, upscaling it to 4k with Spline36 to tax the GPU.

popcornmix commented:
Tried this on Pi 5. Lots of these errors:
```
2023-10-24 18:17:09.130 T:1262    error <general>: failed to create EGL sync object (EGL_BAD_ATTRIBUTE)
2023-10-24 18:17:09.130 T:1262    debug <general>: EGL Debugging:
                                                   Error: EGL_BAD_PARAMETER
                                                   Command: eglDupNativeFenceFDANDROID
                                                   Type: EGL_DEBUG_MSG_ERROR_KHR
                                                   Message: eglDupNativeFenceFDANDROID
```
[kodi.log](http://sprunge.us/U3gjD2)

neo1973 commented:
Looks like Pi 5 is missing `EGL_ANDROID_native_fence_sync`?
```
2023-10-24 18:17:09.012 T:1262     info <general>: EGL_EXTENSIONS = EGL_ANDROID_blob_cache EGL_EXT_buffer_age EGL_EXT_image_dma_buf_import EGL_EXT_image_dma_buf_import_modifiers EGL_KHR_cl_event2 EGL_KHR_config_attribs EGL_KHR_context_flush_control EGL_KHR_create_context EGL_KHR_create_context_no_error EGL_KHR_fence_sync EGL_KHR_get_all_proc_addresses EGL_KHR_gl_colorspace EGL_KHR_gl_renderbuffer_image EGL_KHR_gl_texture_2D_image EGL_KHR_gl_texture_3D_image EGL_KHR_gl_texture_cubemap_image EGL_KHR_image EGL_KHR_image_base EGL_KHR_image_pixmap EGL_KHR_no_config_context EGL_KHR_reusable_sync EGL_KHR_surfaceless_context EGL_EXT_pixel_format_float EGL_KHR_wait_sync EGL_MESA_configless_context EGL_MESA_drm_image EGL_MESA_image_dma_buf_export EGL_MESA_query_driver 
```
Interesting because even on my old Pi 3 it's available:
```
2023-10-24 21:56:53.018 T:28090    info <general>: EGL_EXTENSIONS = EGL_ANDROID_blob_cache EGL_ANDROID_native_fence_sync EGL_EXT_buffer_age EGL_EXT_image_dma_buf_import EGL_EXT_image_dma_buf_import_modifiers EGL_KHR_cl_event2 EGL_KHR_config_attribs EGL_KHR_context_flush_control EGL_KHR_create_context EGL_KHR_create_context_no_error EGL_KHR_fence_sync EGL_KHR_get_all_proc_addresses EGL_KHR_gl_colorspace EGL_KHR_gl_renderbuffer_image EGL_KHR_gl_texture_2D_image EGL_KHR_gl_texture_cubemap_image EGL_KHR_image EGL_KHR_image_base EGL_KHR_image_pixmap EGL_KHR_no_config_context EGL_KHR_reusable_sync EGL_KHR_surfaceless_context EGL_EXT_pixel_format_float EGL_KHR_wait_sync EGL_MESA_configless_context EGL_MESA_drm_image EGL_MESA_image_dma_buf_export EGL_MESA_query_driver EGL_WL_bind_wayland_display
```

lrusak commented:
I've added another commit to change the ordering of the fencing. Looking at kms cube I'm not really sure why I had it different.

chewitt commented:
I've played with this a little on an older Amlogic board (S905) with the current commit. No issues browsing around the GUI or playing media and AFAICT everything remained the same.

smp79 commented:
> I've added another commit to change the ordering of the fencing.

This greatly improved the performance! It is now identical to https://github.com/xbmc/xbmc/pull/23793 and https://github.com/xbmc/xbmc/pull/22012

lrusak committed:
CEGLFence: add support for using EGL_KHR_fence_sync and EGL_ANDROID_native_fence_sync

Signed-off-by: Lukas Rusak <lorusak@gmail.com>

lrusak committed:
CDRMAtomic: add support for using DRM_MODE_ATOMIC_NONBLOCK

Signed-off-by: Lukas Rusak <lorusak@gmail.com>

neo1973 commented on xbmc/windowing/gbm/drm/DRMUtils.h:
```diff
@@ -62,6 +62,15 @@ class CDRMUtils
+  void SetInFenceFd(int fd) { m_inFenceFd = fd; }
+  int GetOutFenceFd()
+  {
+    int fd{-1};
+    std::swap(m_outFenceFd, fd);
+    return fd;
```
```suggestion
    return std::exchange(m_outFenceFd, -1);
```
Also maybe rename the method to something like `TakeOutFenceFd` to emphasis that this alters the object?

lrusak replied:
Great suggestions :+1:

neo1973 reviewed (commented):
I guess then the main thing remaining is some runtime detection if the platform supports `EGL_ANDROID_native_fence_sync`.

lrusak commented:
> I guess then the main thing remaining is some runtime detection if the platform supports `EGL_ANDROID_native_fence_sync`.

This should be solved in the latest commits. I've split and reorganized the commits logically and added runtime detection of the required EGL extensions.

popcornmix commented:
Is there a reason this uses `EGL_ANDROID_native_fence_sync` rather than `EGL_KHR_fence_sync`

My understanding is the latter is the generic, non-platform specific extension, and the former is Android specific (although supported by some other platforms).

The latter is supported by Pi4/5 but not the former.

neo1973 commented:
I don't think plain `EGL_KHR_fence_sync` can convert to or from a file descriptor which are required for the `IN_FENCE_FD` and `OUT_FENCE_FD` DRM properties. Or am I missing something?

popcornmix commented:
You are probably right. I think the function naming is confusing me.
I'll see if I can find out why `EGL_ANDROID_native_fence_sync` is supported by pi3 driver but not by pi4 driver.

lrusak commented:
> You are probably right. I think the function naming is confusing me. I'll see if I can find out why `EGL_ANDROID_native_fence_sync` is supported by pi3 driver but not by pi4 driver.

Looks like VC4 supports the `create_fence_fd` method:

see: https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/include/GL/internal/dri_interface.h#L402-415

VC4: https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/gallium/drivers/vc4/vc4_fence.c
V3D: https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/gallium/drivers/v3d/v3d_fence.c

lrusak commented:
I played with this on i.MX6

Without:
![image](https://github.com/xbmc/xbmc/assets/1616460/f9797e84-c007-4f80-9fd4-e28141b83553)

With:
![image](https://github.com/xbmc/xbmc/assets/1616460/ec46fb0d-3d4a-458e-884f-267e7bd426b1)

popcornmix commented:
Apparently this PR adds support for `EGL_ANDROID_native_fence_sync` on Pi4:
https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/20602

but exposing it caused some other issues for Android. I'll see if I can apply it manually to LE build.

popcornmix commented:
With the mesa PR added, kodi does detect the `EGL_ANDROID_native_fence_sync` on Pi4 and Pi5.

Testing pi4, 1080p60 display (and GUI), and viewing Twelve Monkeys scrolling plot info shows:
30.0 fps current master
60.0 fps with this PR
60.0 fps with [LockFrontBuffer](https://github.com/xbmc/xbmc/pull/22012)

So I can see the gui benefit from blocking less.

On a quick test, video playback using "Direct to plane" and "GL" look okay.

I've also tested without the mesa PR and I get the message
> warning <general>: [GBM] missing support for EGL_KHR_fence_sync and EGL_ANDROID_native_fence_sync - performance may be impacted

and behaviour appears as it was before this PR.

So all fine with current testing.

lrusak commented:
@popcornmix thanks for investigating and testing!

popcornmix commented:
Actually I am seeing an issue with this. After 20 minutes of sitting at initial menu after boot the display locks up.
```
2023-10-30 20:28:10.397 T:1002    error <general>: failed to duplicate EGL fence fd (EGL_SUCCESS)
2023-10-30 20:28:10.431 T:1002    debug <general>: EGL Debugging:
                                                   Error: EGL_BAD_PARAMETER
                                                   Command: eglClientWaitSyncKHR
                                                   Type: EGL_DEBUG_MSG_ERROR_KHR
                                                   Message: _eglClientWaitSyncCommon
2023-10-31 10:42:26.734 T:1002    debug <general>:   EGL_CONFIG_ID: 18
2023-10-31 10:42:26.734 T:1002    debug <general>:   EGL_LEVEL: 0
```
(and the `Error: EGL_BAD_PARAMETER` errors continue indefinitely)
Is it possible there is a leak of duplicated fence fd's? Here is full log:
http://ix.io/4KqM

fritsch commented:
The TakeOutFence method looks suspicious by replacing it with a non initialized version without closing the other fd? It's coming there from the FlushFence method which makes the duplicate?

lrusak commented:
> The TakeOutFence method looks suspicious by replacing it with a non initialized version without closing the other fd? It's coming there from the FlushFence method which makes the duplicate?

```

2. Who is responsible for dup'ing the native fence file descriptors?

    RESOLVED: Whenever a file descriptor is passed into or returned from an
    EGL call in this extension, ownership of that file descriptor is
    transfered. The recipient of the file descriptor must close it when it is
    no longer needed, and the provider of the file descriptor must dup it
    before providing it if they require continued use of the native fence.
```

So I take it that the FD ownership is transferred to EGL.

Ref: https://registry.khronos.org/EGL/extensions/ANDROID/EGL_ANDROID_native_fence_sync.txt

popcornmix commented:
It's possible this is a bug in the mesa side on Pi4. The leak can be spotted quite easily:
```
LibreELEC:~ # while : ; do ls /proc/$(pidof kodi.bin)/fd | wc -l; sleep 1; done
9383
9416
9447
9477
9508
9539
9571
9603
```
The number increased monotonically until it crashes (it is possible other platforms allow a higher number, so may run longer). Can anyone confirm that this number increases monotonically (or not) on a non-pi platform?

Make sure the kodi display has some visible changes (e.g. enable debug overlay with ctrl-shift-d) or the leak stops.

fritsch commented:
Hard to follow honestly. I see the FD duped from internal, but returned to us - and here I cannot follow who should properly closed the duped one.

smp79 commented:
> Can anyone confirm that this number increases monotonically (or not) on a non-pi platform?

```
LibreELEC:~ # while : ; do ls /proc/$(pidof kodi.bin)/fd | wc -l; sleep 1; done
47
46
47
46
46
46
46
46
46
46
46

```
I get a steady 46-47 when debug overlay is active. That's on Intel.

lrusak commented:
> Hard to follow honestly. I see the FD duped from internal, but returned to us - and here I cannot follow who should properly closed the duped one.

The duped FD (from `FlushFence`) get's closed after the drm atomic commit has been posted.

lrusak commented:
I think I need to do a check if the DRM driver supports the required properties. 

Ref: https://drmdb.emersion.fr/properties/4008636142/IN_FENCE_FD

https://drmdb.emersion.fr/properties/3435973836/OUT_FENCE_PTR

popcornmix commented:
Good news. I got a suggested fix for the mesa patch and it appears to work.
Number of file handles stays in 48-51 range.

This is the fixed [patch](https://paste.ubuntu.com/p/Ms5SD6Vwfr/).

fritsch commented:
Awesome. Could you put it into a kodi pastebin? Sadly I don't have an account for the ubuntu pastebin. Want to see how they fixed it.

lrusak commented:
I plan to merge this soon, we can fix any other runtime issues as they come up

kyak commented:
With this PR i see this problem in master:

```
2023-11-20 20:03:07.562 T:3174    error <general>: failed to duplicate EGL fence fd (EGL_SUCCESS)
2023-11-20 20:03:07.568 T:3174     info <general>: Skipped 1 duplicate messages..
2023-11-20 20:03:07.568 T:3174    error <general>: CDRMAtomic::DrmAtomicCommit - atomic commit failed: Device or resource busy
```

With lots of tearing/horizontal line when scrolling.

This is Orange Pi 5 with mesa-panfork.

neo1973 commented:
@kyak: Please open an issue and provide a full debug log there.

heitbaum commented:
- Issue also happened with Intel ADL https://github.com/LibreELEC/LibreELEC.tv/issues/8488
- Revert of patch returns to working system
- This issue occurs when coming out of screensaver

```
2024-01-06 23:44:14.231 T:4971    debug <CWebserver[8080]>: request received for /jsonrpc
2024-01-06 23:44:14.231 T:3789    debug <general>: EGL Debugging:
                                                   Error: EGL_BAD_PARAMETER
                                                   Command: eglClientWaitSyncKHR
                                                   Type: EGL_DEBUG_MSG_ERROR_KHR
                                                   Message: _eglClientWaitSyncCommon
```

neo1973 commented:
Interesting, no such issues on a RPi3B (GLES) or Ryzen 7 5700U (GL and GLES), both running Arch. If you think it's a Kodi issue and not a platform issue please open a GitHub issue and if possible provide a debug log with component logging enabled for _windowing_.

chcore commented:
I raised issue #25222 which looks like it could be this.
